# Chapter 3: Overview

**Discrete Data:** Data can be listed and placed in order - number of heads on 20 coin flips **Can Be Counted**<br />  **Continuous Data:** Can be measured, or take on values in an interval - time of day **Can Be Measured**

**Descriptive Stats:** simply show the stats <br /> **Inferential Stats:** Use samples to make actual inferences from the dataset <br /> **Univariate:** One Variable <br /> **Bivariate:** Two Variables <br /> **Random Variable:** Numerical outcome of a random phenomenon or an experiment

# Chap 4 Univariate Data

**Three Important Aspects of univariate data: Shape, Center, and Spread** <br />
**Shape:** Gaps and clusters, the shape of the distribution (Bell Curve, Uniform, skewed (left/right) <br />
**Center:** Mean, median and Mode <br />
**Spread:** Range, interquartile range, standard deviation

## Types of Graphs

### DotPlot
![](Images/dotplot_ex.png)

![](stem_leaf_plant_img_1.jpg)


### Steam and Leaf Plot
![](Images/stem_leaf_plant_img_1.jpg)

### Histogram
**Bar Graphs are used for Qualitative Data whilst Histograms are used for Quantitative Data**
![](Images/histogram-example-1.png)

### Mean

**x̄** = mean (for statistics (a subset of the population) <br />
**μ(mu)** = Mean for a population (The entire dataset)


x̄ = ( Σ xi ) / n

The term “x̄” represents the sample mean.  The symbol ‘Σ xi’ used in this formula represents the represents the sum of all scores present in the sample (say, in this case) x1 x2 x3 and so on.  The symbol ‘n,’ represents the total number of individuals or observations in the sample.

In [1]:
def mean(a_list):
    return sum(a_list) / len(a_list)

### Median

The Median of an **Ordered** Data set is the "Middle" value in the set

Generally the position of the median can be found using the formula n+1/2, so if you have 28 terms 28+1/2 = 14.5, this is where you'd expect the median <br />
*Note* if the n is odd it will be a whole number 29+1/2 = 15

## Challenge: find the median for Babe Ruth's homerun totals per season

In [4]:
babes_hr = [0,4,3,2,11,29,54,59,35,41,46,25,47,60,54,46,49,46,41,34,22,6]

In [6]:
babes_hr.sort()
print(babes_hr)

[0, 2, 3, 4, 6, 11, 22, 25, 29, 34, 35, 41, 41, 46, 46, 46, 47, 49, 54, 54, 59, 60]


In [7]:
len(babes_hr)

22

In [10]:
median_position = (len(babes_hr)+1) / 2
print(median_position)

11.5


In [15]:
eleven_pos = babes_hr[10]
twelve_pos = babes_hr[11]
median = (twelve_pos + eleven_pos) / 2
print(median)

38.0


### Use mean or median for measure of center; which one?

Depends of the shape of the data: if the distribution is symmetric then the mean and median should be close. However, the median is the better choice if the shape isn't symmetric and more outliers are present.

The Median is superior in these situations because it is a **resistant Statistic**, meaning: it's numerical value isn't dramatically affected by extreme values.

### Variance & Standard Deviation

Vairance is the average squared deviation from the mean. But this won't match the measure of the orginal mean. <br />
If you find the variance of a measurments made in inches, the variance will be in square inches. To correct for this we take the square root of the variance and get **Standard Deviation**

s2 = Σ ( xi – x̄ )2 * 1 / ( n – 1 )

The statistic called sample standard deviation, is a measure of the spread (variability) of the scores in the sample on a given variable and is represented by:

s = sqrt [ Σ ( xi – x̄ )2 * 1/ ( n – 1 ) ]

The term ‘Σ ( xi – x̄ )2’ represents the sum of the squared deviations of the scores from the sample mean.

### Interquartile Range

Although the SD works well in situations where the mean works well (reasonably symmetrical distributions) we need Interquartile range as a measure when we prefer to use the Median

The median essentially splits a data set in two, from here we can find the median of the numbers on either side of the Prime Median. These are called Quartiles. <br />
The median of the lower half is called the lower or first quartile <br />
The median of the upper half is called the upper or third quartile

EX1

In [20]:
# Examples
data_set = [5,5,6,7,8,9,11,13,17]
median_pos = (len(data_set) +1) / 2
median = data_set[4] # list start at index 0 not 1
print(median)

8


In [26]:
lower_half = []
upper_half = []

for num in data_set:
    if num < 8:
        lower_half.append(num)
    elif num > 8:
        upper_half.append(num)
print(f"Lower half: {lower_half}, Upper half: {upper_half}")

q1_pos = (len(lower_half)+1) /2
q2_pos = (len(upper_half)+1) /2
print(f"Q1 Position: {q1_pos} Q2 Position: {q2_pos}")

q1 = (lower_half[1] + lower_half[2]) /2
q2 = (upper_half[1] + upper_half[2]) /2
print(f"Q1: {q1} and Q2: {q2}")

Lower half: [5, 5, 6, 7], Upper half: [9, 11, 13, 17]
Q1 Position: 2.5 Q2 Position: 2.5
Q1: 5.5 and Q2: 12.0


EX2

**Standard Deviation**

In [24]:
babes_hr = [0,4,3,2,11,29,54,59,35,41,46,25,47,60,54,46,49,46,41,34,22,6]

In [25]:
def standard_dev(a_list):
    mean = sum(a_list) / len(a_list)
    x_sum = 0
    
    for num in a_list:
        x_sum += (num - mean)**2
        
    return round((x_sum* 1/(len(a_list)-1))**.5,2)
    

In [26]:
standard_dev(babes_hr)

20.21

**Median**

In [3]:
def median(list1):
    easy_median = False
    median = 0.0
    list1.sort()
    
    median_pos = (len(list1)+1) / 2
    
    if int(median_pos) == median_pos:
        easy_median = True
        
    if easy_median == True:
        median = list1[int(median_pos)-1]
        return float(median)
    else:
        lower_median_pos = int(median_pos-1)
        upper_median_pos = int(median_pos)
        median = ((list1[upper_median_pos]) + (list1[lower_median_pos])) /2
        return float(median)

In [414]:
babes_hr = [0,4,3,2,11,29,54,59,35,41,46,25,47,60,54,46,49,46,41,34,22,6] # Even Number

In [415]:
median(babes_hr)

38.0

**Interquartile Ranges Q1, Q3**

In [4]:
def find_q1_q2(list22):
    low_list = []
    high_list = []
    list22.sort()
    

    median_position = (len(list22)+1) / 2
    for num in list22:
        if (list22.index(num)) +1 < median_position:
            low_list.append(num)

        elif (list22.index(num))+1 > median_position:
            high_list.append(num)
            
    return median(low_list), median(high_list)

In [5]:
babes_hr = [0, 2, 3, 4, 6, 11, 22, 25, 29, 34, 35, 41, 41, 46, 46, 46, 47, 49, 54, 54, 59, 60]
find_q1_q2(babes_hr)

(11.0, 47.0)

### Interquartile range is found by subtracting Quartile 3 from Quartile 1

In [5]:
def interquartile_range(q1,q3):
    
    return q3-q1

In [7]:
interquartile_range(11,47)

36

### Outliers

Most texts define outliers in terms of how far the datapoint is above or below the quartiles of a distribution. To find if a distribution has any outliers, we do the following rule known as **"1.5IQR Rule**

Steps to do this: <br />
Find the IQR <br />
Multiply the IQR by 1.5 <br />
Find Q1-1.5(IQR) and Q3+1.5(IQR) <br />
Any Values below Q1-1.5(IQR) or above Q3+1.5(IQR) is an outlier

**Find Outliers** using the 1.5IQR Rule

In [6]:
def is_outlier_median(a_list):
    
    q1,q3 = find_q1_q2(a_list)
    iqr = q3 - q1
    low_outlier_marker = q1-(1.5*iqr)
    high_outlier_marker = q3+(1.5*iqr)
    
    outlier_list = []
    
    for item in a_list:
        if item < low_outlier_marker:
            outlier_list.append(item)
        elif item > high_outlier_marker:
            outlier_list.append(item)
    
    return outlier_list

In [10]:
book_example = [4.03, 3.76, 3.77, 3.34, 3.47, 2.92, 3.20, 2.71, 3.53, 4.51, 4.56]

In [145]:
is_outlier(book_example)

[]

**Find Outliers** Mean Method (If data point is two or three sdv from mean

In [60]:
def is_outlier_mean(a_list):
    
    mean1 = mean(a_list)
    sdv = standard_dev(a_list)
    outlier_low = []
    outlier_high = []
    
    for num in a_list:
        if num > (mean1+(sdv*2)) and num > (mean1+(sdv*3)):
            outlier_high.append(num)
        elif num < (mean1-(sdv*2)) and num < (mean1-(sdv*3)):
            outlier_high.append(num)
        elif num > (mean1+(sdv*2)):
            outlier_low.append(num)
        elif num < (mean1-(sdv*2)):
            outlier_low.append(num)
        else:
            pass
    
    return(outlier_high, outlier_low)

In [136]:
is_outlier_mean(book_example1)

([], [1])

In [140]:
book_example1 = [1,4.03, 3.76, 3.77, 3.34, 3.47, 2.92, 3.20, 2.71, 3.53, 4.51, 4.56]

In [141]:
give_stats(book_example1)

Median: 3.5
Mean: 3.4
Standard Deviation: 0.94
Q1: 3.06
Q3: 3.9
Interquartile Range: 0.84
Outliers Median Method: [1]
Outliers Mean Method Strong (3+SDV): []
Outliers Mean Method Weak (2SDV): [1]


# Give Stats for a List of Numbers

In [8]:
def give_stats(a_list):
    q1,q3= find_q1_q2(a_list)
    interquartile_range(q1,q3)
    
    print(f"Median: {median(a_list)}")
    print(f"Mean: {round(mean(a_list),2)}")
    print(f"Standard Deviation: {round(standard_dev(a_list),2)}")
    print(f"Q1: {round(find_q1_q2(a_list)[0],2)}")
    print(f"Q3: {round(find_q1_q2(a_list)[1],2)}")
    print(f"Interquartile Range: {round(interquartile_range(q1,q3),2)}")
    print(f"Outliers Median Method: {is_outlier_median(a_list)}")
    print(f"Outliers Mean Method Strong (3+SDV): {is_outlier_mean(a_list)[0]}")
    print(f"Outliers Mean Method Weak (2SDV): {is_outlier_mean(a_list)[1]}")
    print(f"Minimum Value: {min(a_list)}")
    print(f"Maximum Value: {max(a_list)}")

In [11]:
give_stats(book_example)

Median: 3.53
Mean: 3.62
Standard Deviation: 0.59
Q1: 3.2
Q3: 4.03
Interquartile Range: 0.83
Outliers Median Method: []
Outliers Mean Method Strong (3+SDV): []
Outliers Mean Method Weak (2SDV): []
Minimum Value: 2.71
Maximum Value: 4.56


In [142]:
give_stats(babes_hr)

Median: 38.0
Mean: 32.45
Standard Deviation: 20.21
Q1: 11.0
Q3: 47.0
Interquartile Range: 36.0
Outliers Median Method: []
Outliers Mean Method Strong (3+SDV): []
Outliers Mean Method Weak (2SDV): []


## Five Number Summary

Five Number Summary is often used to highlight the Median centric metrics of a dataset <br />
These metrics are: Minimum, Q1, Median, Q3 and Maximum <br />
We can use out Give_Stats Method to get all of this (and More!)

In [12]:
example = [2.8, 3.9, 4.6, 5.3, 10.2, 9.8, 7.7, 13, 2.1, .3, 9.8, 5.3, 9.8, 2.7, 3.9, 7.7, 7.6, 10.1, 8.4, 8.3]

In [13]:
give_stats(example)

Median: 7.65
Mean: 6.67
Standard Deviation: 3.39
Q1: 3.9
Q3: 9.8
Interquartile Range: 5.9
Outliers Median Method: []
Outliers Mean Method Strong (3+SDV): []
Outliers Mean Method Weak (2SDV): []
Minimum Value: 0.3
Maximum Value: 13


### Boxplot (Box and Whisker)

The Boxplot is simply a graphical representation of the 5 number summary <br />
A Box is drawn from q1 to q2 with two "whisker" that extend out to the min and max

![](Images/boxplots.png)

### Percentile Rank of Term

The **Percentile Rank** of a term in a distribution equals the porportion of terms in the distribution that the term is greater than. A term at the 75th percentile is larger than 75% of terms in the distribution. <br />
If we know the 5 number summary then Q1 is at the 25th percentile the median is at the 50th percentile and Q3 is at the 75th percentile.

### Z-Score

One way to identify the position of a term in a distribution is to note how many standard deviations the term is above or below the mean.

In [10]:
def z_score(num, sdv, mean):
    
    return round((num - mean) / sdv,2)
    

In [19]:
z_score(68,3,73)

-1.67

IF a Z-score is positive it is above the mean <br />
If a Z-score is negative is is below the mean

# Normal Distributions

Are symmetric and shaped in a bell like curve: Mean and Standard Deviation are most useful with them

## Empirical Rule

Also known as the **65-95-99.7 Rule** states that approx 68% of the terms in a normal distribution are within one stdv of the mean, 95% are within two stdv of the mean and 99.7% are three stdv within the mean

![](Images/empirical_rule_graph2.png)

In [20]:
z_score(72,3,70)

0.67

## Standard Normal Distribution

Because bell curves are normally distributed, we're often interested in what porportion will lie above or below a given value. <br />
We can do this by using ***Standard Normal Problabilities*** using the z-table below (image isn't entire table)

![](Images/z-table.png)

To find what porportion lies to the left of a given value; first find the z-score for your value

Example: 70 is the mean height, the standard deviation is 3. What porportion lies to the **left** of this value?

In [21]:
z_score(72,3,70)

0.67

Above is our z-score, so we will find .6 on the vertical column, then we will find .07 on the horizantal column, this will give us the porportion to the **left** of our value (and therefore smaller than our value)

In this case about 75% of the terms are below our height value of 72inches

To find how many are to the **right** and therefore greater than our value, simply subtract our value from 100 <br />
100-75 = 25% of terms are greater than our example height of 72 inches.

# Example

How tall must a man be to be in the top 10% of height?

First we do our best to find .89 or .90 on the Z-Table (The book doesn't give a better way :/) <br />
Doing so we find .89 has a Z-score of 1.2. so we can solve for x via the Z-score formula <br />
z = (x-mean)/ standard deviation <br />
in our case 1.2 = (x-70)/3 -> solved x= 73.84 <br />
So a man must be 73.84 inches tall to be in the top 10% of men by height

# Rapid Review

In [23]:
# 7
data_set = [12,13,13,14,16,17,20,28]
give_stats(data_set)

Median: 15.0
Mean: 16.62
Standard Deviation: 5.29
Q1: 13.0
Q3: 18.5
Interquartile Range: 5.5
Outliers Median Method: [28]
28
Outliers Mean Method Strong (3+SDV): []
28
Outliers Mean Method Weak (2SDV): [28]
Minimum Value: 12
Maximum Value: 28


# Free Response

**Question 1**

In [11]:
mantle = [13,23,21,27,37,52,34,42,31,40,54,30,15,35,19,23,22,18]

In [12]:
median(mantle)

28.5

In [13]:
find_q1_q2(mantle)

(21.0, 37.0)

In [14]:
is_outlier_median(mantle)

[]

In [15]:
standard_dev(mantle)

11.94376803070295

In [16]:
mean(mantle)

29.77777777777778

In [17]:
is_outlier_mean(mantle)

54


([], [54])

Yes, 54 would be an outlier via the soft mean outlier rule as outlier > mean + (sdv X 2) <br />
29.7 + (11.9 x 2) = 53.57 and 54 is greater than 53.57

**Question 3**

In [18]:
len(mantle)

18

**Question 4**

In [19]:
scores = [66,82,76,79,72,98,75,80,76,55,77,68,69,84,89]

In [20]:
mean(scores)

76.4

In [27]:
standard_dev(scores)

10.17

In [29]:
z_score(84,10.17,76.4)

0.75

In [30]:
z_score(89,10.17,76.4)

1.24

**Question 5**

In [31]:
give_stats(scores)

Median: 76.0
Mean: 76.4
Standard Deviation: 10.17
Q1: 70.5
Q3: 82.0
Interquartile Range: 11.5
Outliers Median Method: []
98
Outliers Mean Method Strong (3+SDV): []
98
Outliers Mean Method Weak (2SDV): [55, 98]
Minimum Value: 55
Maximum Value: 98


**Question 6**

In [32]:
z_score(.37,.032,.265)

3.28

**Question 7**

In [33]:
m_doubles = [8,11,12,12,14,15,16,17,17,17,21,22,23,24,25,25,28,37]

In [34]:
median(m_doubles)

17.0

In [35]:
find_q1_q2(m_doubles)

(14.5, 24.5)

In [36]:
print(min(m_doubles))
print(max(m_doubles))

8
37


In [38]:
interquartile_range(14.5,24.5)

10.0

**Question 11**

In [39]:
salaries = [6.2,5.4,5.4,4.9,4.4,4.4,3.4,3.3,3,2.4,2.3,1.3,.3,.3]

In [40]:
len(salaries)

14

In [41]:
give_stats(salaries)

Median: 3.3499999999999996
Mean: 3.36
Standard Deviation: 1.88
Q1: 2.3
Q3: 4.9
Interquartile Range: 2.6
Outliers Median Method: []
Outliers Mean Method Strong (3+SDV): []
Outliers Mean Method Weak (2SDV): []
Minimum Value: 0.3
Maximum Value: 6.2


**Question 12**

In [56]:
pennies = []
def multiply(number,marker,lst):
    
    for num in range(0,number+1):
        lst.append(marker)

In [58]:
multiply(163,0,pennies)
multiply(87,1,pennies)
multiply(52,2,pennies)
multiply(75,3,pennies)
multiply(44,4,pennies)
multiply(24,5,pennies)
multiply(36,6,pennies)
multiply(14,7,pennies)
multiply(11,8,pennies)
multiply(5,9,pennies)
multiply(12,10,pennies)
multiply(2,11,pennies)


In [61]:
give_stats(pennies)

Median: 2.0
Mean: 2.48
Standard Deviation: 2.66
Q1: 0.0
Q3: 4.0
Interquartile Range: 4.0
Outliers Median Method: [11, 11, 11]
Outliers Mean Method Strong (3+SDV): [11, 11, 11]
Outliers Mean Method Weak (2SDV): [8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 9, 9, 9, 9, 9, 9, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10]
Minimum Value: 0
Maximum Value: 11


**Question 13**